libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2915,8 +2915,9 @@ bool TypeChecker::visit(IndexAccess const& _access)
 	case Type::Category::TypeType:
 	{
 		TypeType const& typeType = dynamic_cast<TypeType const&>(*baseType);
-		if (dynamic_cast<ContractType const*>(typeType.actualType()))
-			m_errorReporter.typeError(2876_error, _access.location(), "Index access for contracts or libraries is not possible.");
+		if (auto const* contractType = dynamic_cast<ContractType const*>(typeType.actualType()))
+			if (contractType->contractDefinition().isLibrary())
+				m_errorReporter.typeError(2876_error, _access.location(), "Index access for library types and arrays of libraries are not possible.");
```

Comment:
Are arrays of libraries allowed in the first place?

Reply:
Also, why is the error given only for libraries? Before it was given for any contract type

Reply:
``(Index access for library types) and (arrays of libraries) are not possible.``, not ``Index access for (library types and arrays of libraries) are not possible.``.... if you suggest a better message, I'm happy to change it...

Reply:
> Also, why is the error given only for libraries? Before it was given for any contract type

That would be the bug this PR fixes :-). ``C[12];`` is valid just as ``uint[43];`` is. It's just a problem for libraries, since library names have non-zero stack size (their address), which confuses codegen and because there's no real point in these statements anyways, disallowing them for libraries probably seemed easier to fix those issues back then. I looked at the PR that did it - there was never a reason to do it for contracts as well.

But yeah "Index access for contracts" is (weirdly enough) how ``C[4]`` in ``abi.decode(x, (C[4]));`` is parsed - so it's actually an array type in expression position, just a fluke of our AST that that's parsed as "index access".

Reply:
Ah yea, ok